Support modifiers with simple lambda parameters. (#75400)
dotnet/roslyn: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md

This document lists known breaking changes in Roslyn after .NET 9 general release (.NET SDK version 9.0.100) through .NET 10 general release (.NET SDK version 10.0.100).

> ## 'scoped' in a lambda parameter list is now always a modifier.

Reviewer: >'scoped' [](http://example.com/codeflow?start=3&length=8)

Consider using <code>`scoped`</code> rather than 'scoped'.

> ## 'scoped' in a lambda parameter list is now always a modifier.

Author: ```suggestion
## `scoped` in a lambda parameter list is now always a modifier.
```

***Introduced in Visual Studio 2022 version 17.13***

C# 14 introduces the ability to write a lambda with parameter modifiers, without having to specify a parameter type:
https://github.com/dotnet/csharplang/blob/main/proposals/simple-lambda-parameters-with-modifiers.md

As part of this work, a breaking change was accepted where `scoped` will always be treated as a modifier
in a lambda parameter, even where it might have been accepted as a type name in the past.  For example:

```c#

> ref struct @scoped { }
> var v = (scoped scoped s) => { ... };

Author: ```suggestion
var v = (scoped scoped s) => { ... };

ref struct @scoped { }
```

> // Legal if a type named 'scoped' was in scope (for example, from an external library)

Author: ```suggestion
ref struct @scoped { }
```

> var v = (scoped scoped s) => { ... };

Reviewer: Perhaps include a type declaration:
```csharp
var v = (scoped scoped s) => { };

ref struct @scoped { }
```

```

In C# 14 this will be an error as both 'scoped' tokens are treated as modifiers.  The workaround is to

> use '@' in the type name position like so:

Reviewer: Consider using \`scoped\` rather than 'scoped'.

> In C# 14 this will be an error as both 'scoped' tokens are treated as modifiers.  The workaround is to
> use '@' in the type name position like so:

Author: ```suggestion
In C# 14 this will be an error as both `scoped` tokens are treated as modifiers.  The workaround is to
use `@` in the type name position like so:
```

```c#

> // Legal if a type named 'scoped' was in scope (for example, from an external library)
> var v = (scoped @scoped s) => { ... };

Author: ```suggestion
var v = (scoped @scoped s) => { ... };

ref struct @scoped { }
```

> var v = (scoped @scoped s) => { ... };

Reviewer: ```csharp
var v = (scoped @scoped s) => { ... };

ref struct @scoped { }
```
